constructShortestPath.js

```diff
@@ -0,0 +1,93 @@
+/*
+For this approach, we can represent the input data as graph where the cities 
+connecting to each other can be reprsented as nodes
+
+and the roads can be represented as edges. 
+Also by looking at the input, we can represent the graph as undirected and build an adjacency list.
+Thus, for finding the shortest path between two cities, we can find the shorest path the same way we would between nodes in our graph.
+A BFS would be a good approach, since we know that using a BFS, you're guaranteed to find the shorest path, so we can go with that approach. 
+*/
+
+/*
+Question:
+Given 3 inputs: a start city, an end city, and a list of roads, return the shortest path (as an array) from the start city to the end city.
+*/
+
+const { Queue } = require('./utils/queue');
+
+const buildGraph = (nodes) => {
+    const graph = {};
+
+    for (const node of nodes) {
+        const [nodeOne, nodeTwo] = node;
+
+        const containsNodeOne = graph.hasOwnProperty(nodeOne);
+        const containsNodeTwo = graph.hasOwnProperty(nodeTwo);
+
+        if (!containsNodeOne) graph[nodeOne] = [];
+        if (!containsNodeTwo) graph[nodeTwo] = [];
+
+        graph[nodeOne].push(nodeTwo);
+        graph[nodeTwo].push(nodeOne);
+    }
+
+    return graph;
+};
+
+const createPath = (startNode, endNode, prevNodesMap) => {
+    const path = [];
+    let currentNode = endNode;
+
+    while (currentNode !== null) {
+        path.push(currentNode);
+        currentNode = prevNodesMap[currentNode];
+    }
+
+    return path.reverse();
+};
+
+const getShortestPath = (startNode, endNode, nodes) => {
+    const queue = new Queue();
+    const visited = new Set();
+
+    visited.add(startNode);
+    queue.enqueue({ node: startNode, prevNode: null });
+
+    const prevNodesMap = {};
+    const graph = buildGraph(nodes);
+    while (queue.size() > 0) {
+        // Remove node
+        const { node, prevNode } = queue.dequeue();
+
+        // Process node
+        prevNodesMap[node] = prevNode;
+
+        // Get neighbors
+        const neighbors = graph[node];
+        for (const neighbor of neighbors) {
+            if (visited.has(neighbor)) continue;
+
+            visited.add(neighbor);
+            queue.enqueue({ node: neighbor, prevNode: node });
+        }
+    }
+
+    return createPath(startNode, endNode, prevNodesMap);
```

Comment:
P: Why does your algorithm return the SHORTEST path? It seems like you could overwrite values in your `prevNodesMap` and that could cause you to return something that is a true path, but is not the shortest path.

Reply:
It's because when we reach a node for the first time, you are guaranteed that will be the shortest path. Also, the node that previously added the current node, is the ideal node to come from. Since it's the ideal node to come from, there shouldn't be anything overwritten in our object.

Reply:
The statements you said are all correct, and they are critical things to understand with this algorithm, so good job. However, you have not given the correct answer to my question about overwriting values. There is a different reason for why the overwriting is not an issue. See if you can figure it out!

Reply:
The reason why overwriting is not an issue is because of our visited set. The way the algorithm is written is every time we visit a node, it will be the first time and the node previous will always be the ideal node.

Reply:
Exactly, we never visit any nodes twice because of the visited set.